Sources/Hummingbird/Router/Router+validation.swift

```diff
@@ -0,0 +1,111 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Hummingbird server framework project
+//
+// Copyright (c) 2024 the Hummingbird authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See hummingbird/CONTRIBUTORS.txt for the list of Hummingbird authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+import HTTPTypes
+
+extension Router {
+    /// Route description
+    public struct RouteDescription: CustomStringConvertible {
+        /// Route path
+        public let path: RouterPath
+        /// Route method
+        public let method: HTTPRequest.Method
+
+        public var description: String { "\(method) \(path)" }
+    }
+
+    /// List of routes added to router
+    public var routes: [RouteDescription] {
+        let trieValues = self.trie.root.values()
+        return trieValues.flatMap { endpoint in
+            endpoint.value.methods.keys
+                .sorted { $0.rawValue < $1.rawValue }
+                .map { RouteDescription(path: endpoint.path, method: $0) }
+        }
+    }
+
+    /// Validate router
+    ///
+    /// Verify that routes are not clashing
+    public func validate() throws {
+        func matching(routerPath: RouterPath, pathComponents: [String]) -> Bool {
+            guard routerPath.count == pathComponents.count else { return false }
+            for index in 0..<routerPath.count {
+                if case routerPath[index] = pathComponents[index] {
+                    continue
+                }
+                return false
+            }
+            return true
+        }
+        // get trie routes sorted in the way they will be evaluated after building the router
+        let trieValues = self.trie.root.values().sorted { lhs, rhs in
+            let count = min(lhs.path.count, rhs.path.count)
+            for i in 0..<count {
+                if lhs.path[i].priority < rhs.path[i].priority {
+                    return false
+                }
+            }
+            return true
+        }
+        guard trieValues.count > 1 else { return }
+        for index in 1..<trieValues.count {
```

Comment:
I think this is computationally much heavier than it needs to be, though it will catch more routes that overlap.
Let's break down what we want to validate (and why) in a discussion somewhere.

Reply:
I was intending to catch any route that would never be triggered because a higher priority route was triggered instead.

Reply:
Optimised this. It now does the validation at the trie node level which is considerably faster